blog/content/post/2022-05-04-fobnail_provisioning.md

```diff
@@ -0,0 +1,302 @@
+offline integrity measurement verifiers on the USB device (Fobnail Token) and
+attesters running in Dynamically Launched Measured Environments (DLME). It
+allows the Fobnail owner to verify the trustworthiness of the running system
+before performing any sensitive operation. This project was founded by [NlNet
+Foundation](https://nlnet.nl/). More information about the project can be found
+in the [Fobnail documentation](https://fobnail.3mdeb.com/). Also, make sure to
+read other posts related to this project by visiting
+[fobnail](https://blog.3mdeb.com/tags/fobnail/) tag.
+
+# Scope of current phase
+
+This phase is about provisioning Fobnail Token itself. The closing point of that
+process is creating a certificate for Token that can be used later after
+attestation, only if attestation finishes successfully. Before that happens,
+Fobnail Token must verify that Platform Owner can be trusted, and prepare an
+input for that certificate.
+
+## Assumptions and flow of Fobnail provisioning
+
+It is assumed that provisioning is performed in a secure environment, over a
+secure channel, and encryption of data sent between Platform Owner and Fobnail
+Token isn't necessary. After all, Platform Owner is able to issue certificates,
+so there are bigger worries than lack of encryption during provisioning if it
+gets compromised.
+
+Communication with Platform Owner happens through CoAP. Fobnail Token is a
+client, same as during communication with Attester. This is done to keep code
+footprint small enough to fit in nRF52840 dongle. As a consequence, all
+communication is initiated by the Token, so it is responsible for deciding which
+of the three flows (Fobnail provisioning, platform provisioning or attestation)
+it should follow.
+
+First resource that Fobnail Token asks for during its provisioning is a Platform
+Owner certificate chain. It has a [strictly defined format](TBD: link to docs)
+and its root must be provided to the token during compilation. Fobnail validates
+this chain, but because it has neither timekeeping nor (secure) networking
+capabilities, it isn't able to check validity periods or CRLs.
+
+Second request uses POST method and includes CSR (certificate signing request).
+Platform Owner responses to that with a certificate generated from that CSR,
+signed by Platform Owner's private key. This certificate is verified against
+chain received in previous step and saved in non-volatile memory for later use.
+
+![Diagram of Fobnail provisioning flow](/img/Fobnail-flows-fobnail-provisioning.png)
+
+## Fobnail Token's key
+
+There are no good crates for generating RSA keys on Cortex-M, and in our tests
+using universal implementation wasn't able to generate 2048-bit key in 5 hours.
+We also couldn't use ARM TrustZone CryptoCell 310 which is included in nRF52840
+for this purpose because it doesn't have open-sourced libraries. Thus, currently
+Fobnail Token generates Curve25519 key, with help of [Trussed](https://trussed.dev/).
```

Comment:
@krystian-hebel very interesting issues, is there any hope for resolving that? I mean open version of cryptocell or maybe use of newer cryptocell (I assume on newer MCU) which resolve this terrible dependency?

Reply:
Probably yes. There are open versions for other CCs but we have no way of comparing those other than trying to run them on our hardware. They are [hardly documented](https://infocenter.nordicsemi.com/index.jsp?topic=%2Fps_nrf9160%2Fcryptocell.html), you can read how to enable/disable it, but not how it can be actually used, other than "use library from SDK".

Reply:
Let's try to address that when we will write blog post about next phase.